File: SB_Prototype_Pattern92/C++/SB_Prototype_Pattern92/TestCloning.cpp
#include "TestCloning.h"
#include "CloneFactory.h"
#include "Sheep.h"
#include "Dog.h"


//void TestCloning::main(std::vector<std::wstring>& args)
int main()
{

	// Handles routing makeCopy method calls to the
	// right subclasses of Animal

	CloneFactory* animalMaker = new CloneFactory();

	// Creates a new Sheep instance

	Sheep* sally = new Sheep();

	// Creates a clone of Sally and stores it in its own
	// memory location

	Sheep* clonedSheep = static_cast<Sheep*>(animalMaker->getClone(sally));

	// These are exact copies of each other

    std::wcout << sally->toString() << std::endl;

	std::wcout << clonedSheep->toString() << std::endl;

	std::wcout << L"Sally HashCode: " << &sally << std::endl;

	std::wcout << L"Clone HashCode: " <<  &clonedSheep << "\n" << std::endl;

	// Creates a new Dog instance

	Dog* Sophie = new Dog();

	// Creates a clone of Sally and stores it in its own
	// memory location

	Dog* clonedDog = static_cast<Dog*>(animalMaker->getClone(Sophie));

	// These are exact copies of each other

	std::wcout << Sophie->toString() << std::endl;

	std::wcout << clonedDog->toString() << std::endl;

	std::wcout << L"Sophie HashCode: " << &Sophie << std::endl;

	std::wcout << L"Clone HashCode: " << &clonedDog << std::endl;

	delete animalMaker;
}

	

File: SB_Prototype_Pattern92/java/SB_Prototype_Java92/src/com/company/TestCloning.java
package com.company;

public class TestCloning {

    public static void main(String[] args){

        // Handles routing makeCopy method calls to the
        // right subclasses of Animal

        CloneFactory animalMaker = new CloneFactory();

        // Creates a new Sheep instance

        Sheep sally = new Sheep();

        // Creates a clone of Sally and stores it in its own
        // memory location

        Sheep clonedSheep = (Sheep) animalMaker.getClone(sally);

        // These are exact copies of each other

        System.out.println(sally);

        System.out.println(clonedSheep);

        System.out.println("Sally HashCode: " + System.identityHashCode(System.identityHashCode(sally)));

        System.out.println("Clone HashCode: " + System.identityHashCode(System.identityHashCode(clonedSheep)) + "\n");

        // Creates a new Dog instance

        Dog Sophie = new Dog();

        // Creates a clone of Sally and stores it in its own
        // memory location

        Dog clonedDog = (Dog) animalMaker.getClone(Sophie);

        // These are exact copies of each other

        System.out.println(Sophie);

        System.out.println(clonedDog);

        System.out.println("Sophie HashCode: " + System.identityHashCode(System.identityHashCode(Sophie)));

        System.out.println("Clone HashCode: " + System.identityHashCode(System.identityHashCode(clonedDog)) + "\n");
    }

}
